mne/time_frequency/spectrum.py

```diff
@@ -1179,21 +1214,60 @@ def __getitem__(self, item):
         return BaseRaw._getitem(self, item, return_times=False)
 
 
-def _check_data_shape(data, freqs, info, ndim):
-    if data.ndim != ndim:
-        raise ValueError(f"Data must be a {ndim}D array.")
+def _check_data_shape(data, info, freqs, dimnames, weights):
+    if data.ndim != len(dimnames):
+        raise ValueError(
+            f"Expected data to have {len(dimnames)} dimensions, got {data.ndim}."
+        )
+
+    is_epoched = 1 if "epoch" in dimnames else 0
+    allowed_dims = ["epoch", "channel", "freq", "segment", "taper"]
+    allowed_dims = allowed_dims[0 if is_epoched else 1 :]
+    if set(allowed_dims).intersection(dimnames) != set(dimnames):
+        raise ValueError(
+            f"All entries of `dimnames` must be in {allowed_dims}, got {dimnames}."
+        )
```

Comment:
2 changes here: first, `is_epoched` should be passed in as a bool from the calling function (depending on whether it's the EpochsSpectrumArray constructor, or the (non-epochs) SpectrumArray constructor --- don't rely on whether the user included in `dimnames`.  Second: let's reuse our standard `_check_option` function for checking the dimname entries (unless there's a compelling reason not to that I'm missing?)

```suggestion
    allowed_dims = {"epoch", "channel", "freq", "segment", "taper"}
    if not is_epoched:
        allowed_dims.remove("epoch")
    # TODO maybe we should be nice and allow plural versions of each dimname?
    for dim in dimnames:
        _check_option("dimnames", dim, allowed_dims)
```